mesonbuild/backend/xcodebackend.py

```diff
@@ -226,8 +226,7 @@ def gen_id(self) -> str:
         return str(uuid.uuid4()).upper().replace('-', '')[:24]
 
     def get_target_dir(self, target):
-        dirname = os.path.join(target.get_subdir(), T.cast('str', self.environment.coredata.get_option(OptionKey('buildtype'))))
-        #os.makedirs(os.path.join(self.environment.get_build_dir(), dirname), exist_ok=True)
+        dirname = T.cast('str', self.environment.coredata.get_option(OptionKey('buildtype')))
```

Comment:
I am pretty skeptical of this change.

```
This replicates Xcode's default behavior, in which all build
products are in the exact same directory. Because of this, forcing
the swift compiler to search recursively within subdirectories is
unnecessary and can be removed.
```

This corresponds to the `--layout=flat` option which was deprecated in #9243. I am generally unconvinced this is a good idea and there are a variety of ways it can be broken by projects with expectations.

Is there any reason to do this other than making it "unnecessary" for the swift compiler to do searching? Does this fix something that is broken? And if so, is there another way we can fix that thing without switching to layout=flat?

Reply:
By default, Xcode will generate subdirectories for each target in the project, but it only applies to the source files. It doesn't apply to the generated build products, which will all be in the same directory. This negates the need to have to search that directory recursively.

To get an idea of what I'm talking about, without flattening the build product directory and leaving the swift compiler import flag, here is what the compile command looked like for swift test case 3 on the first attempt:
```
SwiftCompile normal arm64 /Users/alberttang/Repositories/meson/test\ cases/swift/3\ library/exe/main.swift (in target '9095bdb@@dataprog@exe' from project 'linking')
    cd /Users/alberttang/Repositories/meson/test\ cases/swift/3\ library
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swift-frontend -c -primary-file /Users/alberttang/Repositories/meson/test\ cases/swift/3\ library/exe/main.swift -emit-dependencies-path /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/Objects-normal/arm64/main.d -emit-reference-dependencies-path /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/Objects-normal/arm64/main.swiftdeps -serialize-diagnostics-path /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/Objects-normal/arm64/main.dia -target arm64-apple-macos13.3 -Xllvm -aarch64-use-tbi -enable-objc-interop -stack-check -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk -I /Users/alberttang/Repositories/meson/builddir/lib/debug/libDataSource.a.p -I /Users/alberttang/Repositories/meson/builddir/exe/debug -I /Users/alberttang/Repositories/meson/builddir -I /Users/alberttang/Repositories/meson/builddir/build -I /Users/alberttang/Repositories/meson/builddir/exe -I /Users/alberttang/Repositories/meson/builddir/lib -I /Users/alberttang/Repositories/meson/builddir/meson-info -I /Users/alberttang/Repositories/meson/builddir/meson-logs -I /Users/alberttang/Repositories/meson/builddir/meson-private -F /Users/alberttang/Repositories/meson/builddir/exe/debug -no-color-diagnostics -g -swift-version 5 -enforce-exclusivity\=checked -O -serialize-debugging-options -enable-bare-slash-regex -empty-abi-descriptor -Xcc -working-directory -Xcc /Users/alberttang/Repositories/meson/test\ cases/swift/3\ library -resource-dir /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift -Xcc -ivfsstatcache -Xcc /var/folders/ym/0by8c_vx1tj_mx7l88q18yqc0000gn/C/com.apple.DeveloperTools/14.3.1-14E300c/Xcode/SDKStatCaches.noindex/macosx13.3-22E245-.sdkstatcache -Xcc -I/Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/swift-overrides.hmap -Xcc -I/Users/alberttang/Repositories/meson/builddir/exe/debug/include -Xcc -I/Users/alberttang/Repositories/meson/builddir/exe/debug/dataprog.p -Xcc -I/Users/alberttang/Repositories/meson/builddir/exe -Xcc -I/Users/alberttang/Repositories/meson/test\ cases/swift/3\ library/exe -Xcc -I/Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/DerivedSources-normal/arm64 -Xcc -I/Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/DerivedSources/arm64 -Xcc -I/Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/DerivedSources -module-name dataprog -frontend-parseable-output -disable-clang-spi -target-sdk-version 13.3 -target-sdk-name macosx13.3 -o /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/Objects-normal/arm64/main.o -index-unit-output-path /linking.build/debug/9095bdb@@dataprog@exe.build/Objects-normal/arm64/main.o
/Users/alberttang/Repositories/meson/test cases/swift/3 library/exe/main.swift:1:8: error: no such module 'DataSource'
import DataSource
       ^
```

Here is what it looked like on the second attempt:
```
SwiftCompile normal arm64 /Users/alberttang/Repositories/meson/test\ cases/swift/3\ library/exe/main.swift (in target '9095bdb@@dataprog@exe' from project 'linking')
    cd /Users/alberttang/Repositories/meson/test\ cases/swift/3\ library
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swift-frontend -c -primary-file /Users/alberttang/Repositories/meson/test\ cases/swift/3\ library/exe/main.swift -emit-dependencies-path /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/Objects-normal/arm64/main.d -emit-reference-dependencies-path /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/Objects-normal/arm64/main.swiftdeps -serialize-diagnostics-path /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/Objects-normal/arm64/main.dia -target arm64-apple-macos13.3 -Xllvm -aarch64-use-tbi -enable-objc-interop -stack-check -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk -I /Users/alberttang/Repositories/meson/builddir/lib/debug/libDataSource.a.p -I /Users/alberttang/Repositories/meson/builddir/exe/debug -I /Users/alberttang/Repositories/meson/builddir -I /Users/alberttang/Repositories/meson/builddir/EagerLinkingTBDs -I /Users/alberttang/Repositories/meson/builddir/build -I /Users/alberttang/Repositories/meson/builddir/debug -I /Users/alberttang/Repositories/meson/builddir/exe -I /Users/alberttang/Repositories/meson/builddir/lib -I /Users/alberttang/Repositories/meson/builddir/linking.build -I /Users/alberttang/Repositories/meson/builddir/meson-info -I /Users/alberttang/Repositories/meson/builddir/meson-logs -I /Users/alberttang/Repositories/meson/builddir/meson-private -I /Users/alberttang/Repositories/meson/builddir/EagerLinkingTBDs/debug -I /Users/alberttang/Repositories/meson/builddir/build/XCBuildData -I /Users/alberttang/Repositories/meson/builddir/exe/EagerLinkingTBDs -I /Users/alberttang/Repositories/meson/builddir/exe/debug -I /Users/alberttang/Repositories/meson/builddir/exe/linking.build -I /Users/alberttang/Repositories/meson/builddir/lib/EagerLinkingTBDs -I /Users/alberttang/Repositories/meson/builddir/lib/debug -I /Users/alberttang/Repositories/meson/builddir/lib/linking.build -I /Users/alberttang/Repositories/meson/builddir/linking.build/debug -I /Users/alberttang/Repositories/meson/builddir/build/XCBuildData/51c890c9b4b079b6d35ce8121c1c76c2.xcbuilddata -I /Users/alberttang/Repositories/meson/builddir/exe/EagerLinkingTBDs/debug -I /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug -I /Users/alberttang/Repositories/meson/builddir/lib/EagerLinkingTBDs/debug -I /Users/alberttang/Repositories/meson/builddir/lib/debug/DataSource.swiftmodule -I /Users/alberttang/Repositories/meson/builddir/lib/linking.build/debug -I /Users/alberttang/Repositories/meson/builddir/linking.build/debug/REGENERATE.build -I /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build -I /Users/alberttang/Repositories/meson/builddir/lib/debug/DataSource.swiftmodule/Project -I /Users/alberttang/Repositories/meson/builddir/lib/linking.build/debug/76b5a35@@DataSource@sta.build -I /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/DerivedSources -I /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/Objects-normal -I /Users/alberttang/Repositories/meson/builddir/lib/linking.build/debug/76b5a35@@DataSource@sta.build/DerivedSources -I /Users/alberttang/Repositories/meson/builddir/lib/linking.build/debug/76b5a35@@DataSource@sta.build/Objects-normal -I /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/Objects-normal/arm64 -I /Users/alberttang/Repositories/meson/builddir/lib/linking.build/debug/76b5a35@@DataSource@sta.build/Objects-normal/arm64 -F /Users/alberttang/Repositories/meson/builddir/exe/debug -no-color-diagnostics -g -swift-version 5 -enforce-exclusivity\=checked -O -serialize-debugging-options -enable-bare-slash-regex -empty-abi-descriptor -Xcc -working-directory -Xcc /Users/alberttang/Repositories/meson/test\ cases/swift/3\ library -resource-dir /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift -Xcc -ivfsstatcache -Xcc /var/folders/ym/0by8c_vx1tj_mx7l88q18yqc0000gn/C/com.apple.DeveloperTools/14.3.1-14E300c/Xcode/SDKStatCaches.noindex/macosx13.3-22E245-.sdkstatcache -Xcc -I/Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/swift-overrides.hmap -Xcc -I/Users/alberttang/Repositories/meson/builddir/exe/debug/include -Xcc -I/Users/alberttang/Repositories/meson/builddir/exe/debug/dataprog.p -Xcc -I/Users/alberttang/Repositories/meson/builddir/exe -Xcc -I/Users/alberttang/Repositories/meson/test\ cases/swift/3\ library/exe -Xcc -I/Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/DerivedSources-normal/arm64 -Xcc -I/Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/DerivedSources/arm64 -Xcc -I/Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/DerivedSources -module-name dataprog -frontend-parseable-output -disable-clang-spi -target-sdk-version 13.3 -target-sdk-name macosx13.3 -o /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/Objects-normal/arm64/main.o -index-unit-output-path /linking.build/debug/9095bdb@@dataprog@exe.build/Objects-normal/arm64/main.o
```

Paste both commands into a text editor and turn on word wrap. You will then see the problem right away.

Reply:
After a while, I used wdiff and found the difference you refer to:


> -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk -I /Users/alberttang/Repositories/meson/builddir/lib/debug/libDataSource.a.p -I /Users/alberttang/Repositories/meson/builddir/exe/debug -I /Users/alberttang/Repositories/meson/builddir -I {+/Users/alberttang/Repositories/meson/builddir/EagerLinkingTBDs -I+} /Users/alberttang/Repositories/meson/builddir/build -I {+/Users/alberttang/Repositories/meson/builddir/debug -I+} /Users/alberttang/Repositories/meson/builddir/exe -I /Users/alberttang/Repositories/meson/builddir/lib -I {+/Users/alberttang/Repositories/meson/builddir/linking.build -I+} /Users/alberttang/Repositories/meson/builddir/meson-info -I /Users/alberttang/Repositories/meson/builddir/meson-logs -I /Users/alberttang/Repositories/meson/builddir/meson-private {+-I /Users/alberttang/Repositories/meson/builddir/EagerLinkingTBDs/debug -I /Users/alberttang/Repositories/meson/builddir/build/XCBuildData -I /Users/alberttang/Repositories/meson/builddir/exe/EagerLinkingTBDs -I /Users/alberttang/Repositories/meson/builddir/exe/debug -I /Users/alberttang/Repositories/meson/builddir/exe/linking.build -I /Users/alberttang/Repositories/meson/builddir/lib/EagerLinkingTBDs -I /Users/alberttang/Repositories/meson/builddir/lib/debug -I /Users/alberttang/Repositories/meson/builddir/lib/linking.build -I /Users/alberttang/Repositories/meson/builddir/linking.build/debug -I /Users/alberttang/Repositories/meson/builddir/build/XCBuildData/51c890c9b4b079b6d35ce8121c1c76c2.xcbuilddata -I /Users/alberttang/Repositories/meson/builddir/exe/EagerLinkingTBDs/debug -I /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug -I /Users/alberttang/Repositories/meson/builddir/lib/EagerLinkingTBDs/debug -I /Users/alberttang/Repositories/meson/builddir/lib/debug/DataSource.swiftmodule -I /Users/alberttang/Repositories/meson/builddir/lib/linking.build/debug -I /Users/alberttang/Repositories/meson/builddir/linking.build/debug/REGENERATE.build -I /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build -I /Users/alberttang/Repositories/meson/builddir/lib/debug/DataSource.swiftmodule/Project -I /Users/alberttang/Repositories/meson/builddir/lib/linking.build/debug/76b5a35@@DataSource@sta.build -I /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/DerivedSources -I /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/Objects-normal -I /Users/alberttang/Repositories/meson/builddir/lib/linking.build/debug/76b5a35@@DataSource@sta.build/DerivedSources -I /Users/alberttang/Repositories/meson/builddir/lib/linking.build/debug/76b5a35@@DataSource@sta.build/Objects-normal -I /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/Objects-normal/arm64 -I /Users/alberttang/Repositories/meson/builddir/lib/linking.build/debug/76b5a35@@DataSource@sta.build/Objects-normal/arm64+} -F /Users/alberttang/Repositories/meson/builddir/exe/debug -no-color-diagnostics -g -swift-version 5

With particular note to:
```
{+/Users/alberttang/Repositories/meson/builddir/EagerLinkingTBDs -I+}

{+/Users/alberttang/Repositories/meson/builddir/debug -I+}

{+/Users/alberttang/Repositories/meson/builddir/linking.build -I+}

{+-I /Users/alberttang/Repositories/meson/builddir/EagerLinkingTBDs/debug -I /Users/alberttang/Repositories/meson/builddir/build/XCBuildData -I /Users/alberttang/Repositories/meson/builddir/exe/EagerLinkingTBDs -I /Users/alberttang/Repositories/meson/builddir/exe/debug -I /Users/alberttang/Repositories/meson/builddir/exe/linking.build -I /Users/alberttang/Repositories/meson/builddir/lib/EagerLinkingTBDs -I /Users/alberttang/Repositories/meson/builddir/lib/debug -I /Users/alberttang/Repositories/meson/builddir/lib/linking.build -I /Users/alberttang/Repositories/meson/builddir/linking.build/debug -I /Users/alberttang/Repositories/meson/builddir/build/XCBuildData/51c890c9b4b079b6d35ce8121c1c76c2.xcbuilddata -I /Users/alberttang/Repositories/meson/builddir/exe/EagerLinkingTBDs/debug -I /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug -I /Users/alberttang/Repositories/meson/builddir/lib/EagerLinkingTBDs/debug -I /Users/alberttang/Repositories/meson/builddir/lib/debug/DataSource.swiftmodule -I /Users/alberttang/Repositories/meson/builddir/lib/linking.build/debug -I /Users/alberttang/Repositories/meson/builddir/linking.build/debug/REGENERATE.build -I /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build -I /Users/alberttang/Repositories/meson/builddir/lib/debug/DataSource.swiftmodule/Project -I /Users/alberttang/Repositories/meson/builddir/lib/linking.build/debug/76b5a35@@DataSource@sta.build -I /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/DerivedSources -I /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/Objects-normal -I /Users/alberttang/Repositories/meson/builddir/lib/linking.build/debug/76b5a35@@DataSource@sta.build/DerivedSources -I /Users/alberttang/Repositories/meson/builddir/lib/linking.build/debug/76b5a35@@DataSource@sta.build/Objects-normal -I /Users/alberttang/Repositories/meson/builddir/exe/linking.build/debug/9095bdb@@dataprog@exe.build/Objects-normal/arm64 -I /Users/alberttang/Repositories/meson/builddir/lib/linking.build/debug/76b5a35@@DataSource@sta.build/Objects-normal/arm64+}
```

What's not clear to me is, what the problem with this is.

Is the problem that the compiler command lines are long and awkward? While that may be inconvenient, this is not directly user-visible, right?

I have very little knowledge of what all these directories do.

Reply:
The problem is the subdirectories containing the needed build products have to be searched recursively to get a successful build, but they are only picked up on the second build attempt. To better explain this, I'll go a different way.

Let's say we have two targets: `target1` and `target2`. `target1` depends on `target2` to build successfully.

Xcode's default behavior looks like this for each project:
```
build
  - Intermediates.noindex
    - project.build
      - <build configuration>
        - target1.build
          - DerivedSources
          - Objects-normal
        - target2.build
          - (same as target1.build)
  - Products
    - <build configuration> 
      - target1
      - target2.swiftmodule
      - libtarget2.a
```
For the dependencies, Xcode will search `build/Products/<build configuration>`. Recursive searching is not needed here, as no subdirectories are generated.

This is meson's current behavior when subdirectories are involved:
```
build
  - <build configuration>
  - subdir1
    - <build configuration>
      - target1
    - project.build
      - <build configuration>
        - target1.build
          - DerivedSources
          - Objects-normal
  - subdir2
    - <build configuration>
      - target2.swiftmodule
      - libtarget2.a
    - project.build
      - <build configuration>
        - target2.build
          - (same as target1.build)
  - project.build
    - <build configuration>
      - REGENERATE.build
```
The problem here is there are two build products paths, but each target can only have one. That means the dependencies for `target1` must be in `build/subdir1/<build configuration>`, but they're not, so it fails.

Making the swift compiler search the entire build directory recursively seemed like a good solution at first, but I later found that this only works after the first failed attempt. That's why I went with trying to replicate Xcode's default behavior instead, which results in this:
```
build
  - <build configuration>
    - target1
    - target2.swiftmodule
    - libtarget2.a
  - project.build
    - <build configuration>
      - target1.build
        - DerivedSources
        - Objects-normal
      - target2.build
      - (same as target1.build)
      - REGENERATE.build
```
This is where the change you're skeptical of is needed. `target1` now compiles correctly, but the linker fails because it thinks `libtarget2.a` is in `build/subdir2/<build configuration>`, so the change corrects that to `build/<build configuration>`.

Reply:
I am reverting this change. I have a better fix for handling Swift dependencies that does not require changing the build directory structure or scanning it recursively.